### ISSUE TITLE
`getStrLn` uninterruptible

### ISSUE THREAD

**Issue by mperucca** [zio/zio#780]
In case this is unknown, the following never terminates (at least on Windows/JVM) `getStrLn.fork.flatMap(_.interrupt)`.

`blocking(getStrLn).fork.flatMap(f => ZIO.sleep(1.second) *> f.interrupt)` does not terminate either.

I've wanted this to work in order to do `longRunningComputation race getStrLn` to allow cancellation from the user through the command line.

Perhaps it's just my terminal, but I've gotten around this by polling with `scala.Console.in.ready()` while looping `scala.Console.in.read(charBuffer)` to read the input. This seems to interrupt as expected.

P.S. Thanks for an awesome library! It provides so many abstractions I've longed for.

**Comment by jdegoes** [comment]
@mperucca You are right, this needs to be fixed! Thanks for reporting, and also for being an early user. 😄

**Comment by jdegoes** [comment]
To fix this issue, we will have to split apart `console` into separate JVM and Scala.js packages. Then the JVM implementation of `console` will have to use the `blocking` module for `getStrLn`.

**Comment by denisgarci** [comment]
I would like to take a crack at this issue. I have a few questions. Is this the right signature for the new console service on the JVM?
```
object Console extends Serializable {
  trait Service[R] {
    def putStr(line: String): ZIO[R, Nothing, Unit]

    def putStrLn(line: String): ZIO[R, Nothing, Unit]

    val getStrLn: ZIO[R with Blocking, IOException, String]

  }
}
```
Or shoud it still be  `getStrLn: ZIO[R, IOException, String]`?

**Comment by mperucca** [comment]
> Then the JVM implementation of `console` will have to use the `blocking` module for `getStrLn`.

Maybe I'm misunderstanding, but I don't think this will work.

The following doesn't terminate:
`effectBlocking(scala.io.StdIn.readLine()).fork.flatMap(f => ZIO.sleep(1.second) *> f.interrupt)`

What I had to do was to poll the input with `scala.Console.in.ready()` and only read whenever that returned true. I'm not sure how the terminal input is buffered, but it might be necessary to only read with https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#read-char:A-int-int- through `scala.Console.in` instead of `scala.io.StdIn.readLine()` so as to not block if there is no newline character ready to be read.

**Comment by denisgarci** [comment]
@mperucca You're right I tried the `blocking` approach in a local branch and the second examples doesn't work.
```
  def getStrInterruptible1 =
    unsafeRun(
      for {
        input1      <- Live.console.getStrLn.fork.flatMap(_.interrupt)
      } yield input1.interrupted must_=== true
    )

  def getStrInterruptible2 =
    unsafeRun(
      for {
        input1      <- Live.console.getStrLn.fork.flatMap(f => ZIO.sleep(1.second) *> f.interrupt)
      } yield input1.interrupted must_=== true
    )
```
The test `getStrInterruptible1` passes but `getStrInterruptible2` times out.

**Comment by jdegoes** [comment]
@denisgarci 

> Or shoud it still be getStrLn: ZIO[R, IOException, String]?

Yes, it should remain this way.

> The following doesn't terminate:
> `effectBlocking(scala.io.StdIn.readLine()).fork.flatMap(f => ZIO.sleep(1.second) *> f.interrupt)`

You never know with JVM methods! However, we will still need `Blocking` because we will have stuff to run in the unbounded thread pool—we can't choke up the async thread pool with this blocking IO.

I'd suggest an implementation like this:

```scala
trait Live extends Blocking {
  ...
  val getStrLn: IO[IOException, String] = 
    ZIO.uninterruptibleMask(restore => for {
      fiber <- blocking.blocking(
            IO(scala.io.StdIn.readLine()).refineOrDie { 
              case io : IOException => io 
            }).fork
      exit <- restore(fiber.await).onInterrupt(fiber.interrupt)
      line <- IO.done(exit)
    } yield line)
}
```

That's for the JVM. The Scala.js implementation won't have `getStrLn`.

Probably mistakes in the above ^^^

**Comment by mperucca** [comment]
That implementation still does not allow interrupting `getStrLn`. The problem is that `StdIn.readline()` does not acknowledge `Thread.interrupt()`.

**Comment by mperucca** [comment]
I did something like this to allow interruption.
```
def getStrLn: UIO[String] = {
  for {
    buffer <- Ref.make(new StringBuilder())
    line <- {
      def readUntilNewLine(): UIO[String] = UIO.suspend {
        if (scala.Console.in.ready()) {
          val read = scala.Console.in.read()
          if (read == '\n' || read == -1)
            buffer.get.map(_.mkString)
          else
            buffer.update(_ += read.toChar) *> readUntilNewLine()
        } else
          readUntilNewLine()
      }
      readUntilNewLine()
    }
  } yield line
}
```
It has issues though. Just to list a few:
1. It ignores possible `IOException` in `ready` and `read`.
2. It busy polls, maybe use `sleep` or `yieldNow` on iterations when nothing's ready to be read?
3. It doesn't consider line feeds as part of the newline.
4. Concurrent calls to `getStrLn` could still block the underlying thread and not allow interruption.

And there are probably other problems due to me not being familiar enough with ZIO and/or using it incorrectly.

**Comment by jdegoes** [comment]
> That implementation still does not allow interrupting getStrLn. The problem is that StdIn.readline() does not acknowledge Thread.interrupt().

Sorry, my code was wrong. I deleted the most important part while refactoring. 😂 

If we close the same input stream, it will throw `IOException`, which will terminate that fiber. 

```scala
trait Live extends Blocking {
  ...
  val getStrLn: IO[IOException, String] = 
    ZIO.uninterruptibleMask(restore => for {
      fiber <- blocking.blocking(
            IO(scala.Console.in.readLine()).refineOrDie { 
              case io : IOException => io 
            }).fork
      line <- restore(fiber.join).onInterrupt(Task(scala.Console.in.close).option *> fiber.interrupt)
    } yield line)
}
```

Now when the `fiber.join` is interrupted, it closes the input stream, which will throw `IOException` from `readLine`.

> And there are probably other problems due to me not being familiar enough with ZIO and/or using it incorrectly.

It is a good implementation! The main problem is the busy-waiting. It will suck up a lot of CPU.

Like you mention, we can add a sleep there, but then we will introduce latency proportional to the sleep.

_Update_: It looks like standard input close may not result in `readLine` throwing an exception (it may be synchronized to signal after the `readLine` executes). The JRuby community [found a workaround](https://github.com/elastic/logstash/issues/1769) but its applicability will have to be explored. 

If the above investigation fails, it looks like polling is the best we can do.

If we have to do polling, we should `Clock.Live.sleep` for increasing time, say starting at 10ms and eventually increasing to 200ms (but resetting this every time we get another byte). This way it will be low latency in the common case and we only have 1/5th a second of latency in the worse case.

**Comment by mperucca** [comment]
We wouldn't want interrupting `getStrLn` to close standard input though, would we? Although allowing `getStrLn` again after an interrupt brings up the question of what line to return if a previous line got interrupted partway through being read.

**Comment by jdegoes** [comment]
> We wouldn't want interrupting getStrLn to close standard input though, would we?

You tell me. 😄 In a console application it would be pretty unusual to interrupt the main read loop unless the application were terminating, in which case closing standard input would be acceptable.

With this use case: `longRunningComputation race getStrLn`, however, we should not close standard input or else the pattern cannot be repeated....

That suggests to me we investigate the JRuby solution or go with polling.

**Comment by adamgfraser** [comment]
I am working on this.

**Comment by alexknvl** [comment]
Here is a POC for the JRuby approach. I verified that it has the expected behavior on JRE8 + MacOS.

```scala
  @silent def interruptibleStdin(): FileInputStream = {
    try {
      var stdin = System.in
      // we expect System.in to be a FilterInputStream, let's setup access to its "in" field
      val inField = classOf[FilterInputStream].getDeclaredField("in")
      inField.setAccessible(true)
      // normally the underlying FileInputStream is directly in the "in" field but
      // let's make sure it's not hidden in another inner FilterInputStream
      while (stdin.isInstanceOf[FilterInputStream])
        stdin = inField.get(stdin).asInstanceOf[InputStream]
      // no luck discovering inner FileInputStream ?
      if (!stdin.isInstanceOf[FileInputStream]) sys.error("blah")
      stdin.asInstanceOf[FileInputStream]
    } catch {
      case e@(_: NoSuchFieldException | _: IllegalAccessException) =>
        sys.error("blah")
    }
  }

  @silent def interruptibleReader(): BufferedReader = {
    val is = interruptibleStdin()
    val ch = is.getChannel
    assert(ch.isInstanceOf[InterruptibleChannel])

    new BufferedReader(new InputStreamReader(new InputStream {
      override def read(): Int = sys.error("shouldn't be called")

      override def read(b: Array[Byte], off: Int, len: Int): Int = {
        val data = ByteBuffer.allocate(len)
        val n = ch.read(data)
        if (n >= 0) {
          data.position(0)
          data.get(b, off, n)
          n
        }
        else throw new IOException()
      }
    }))
  }

  @silent final val getStrLn =
    zio.blocking.effectBlockingInterrupt { interruptibleReader().readLine() }

  import zio.duration._

  override def run(args: List[String]): ZIO[zio.ZEnv, Nothing, Int] = {
    import zio.console.putStrLn
    val a = getStrLn.flatMap(putStrLn(_))
    val b = ZIO.sleep(10.second) *> putStrLn("No input")
    (a race b).orDie.as(0)
}
```

**Comment by alexknvl** [comment]
Is the usage of reflection a problem? I don't know much about Java 9+.

**Comment by alexknvl** [comment]
An improved POC that shows that both polling and JRuby solution work. Polling can be used as a fallback solution.

<details>
<summary>Snippet</summary>

```scala
  @silent def stdInStream(): Option[FileInputStream] = {
    try {
      var stdin = System.in

      // We expect System.in to be a FilterInputStream,
      // let's setup access to its `in` field.
      val inField = classOf[FilterInputStream].getDeclaredField("in")
      inField.setAccessible(true)

      // Normally the underlying FileInputStream is directly
      // in the "in" field but here we make sure it's not hidden
      // in yet another inner `FilterInputStream`.
      while (stdin.isInstanceOf[FilterInputStream])
        stdin = inField.get(stdin).asInstanceOf[InputStream]

      // The resulting input stream should have type FileInputStream.
      stdin match {
        case fis: FileInputStream =>
          Some(fis)

        case _ =>
          // UNEXPECTED: System.in does not have an underlying FileInputStream.
          None
      }
    } catch {
      case e@(_: NoSuchFieldException | _: IllegalAccessException | _: SecurityException) =>
        // UNEXPECTED: Could not access FilterInputStream's `in` field.
        None
    }
  }

  @silent def interruptibleStdInReader(): () => String = {
    val interruptibleInputStream = stdInStream() match {
      case Some(fis) if false =>
        val channel = fis.getChannel

        new InputStream {
          override def read(): Int =
            sys.error("Unexpected operation performed by BufferedReader")
          override def available(): Int =
            fis.available()

          override def read(b: Array[Byte], off: Int, len: Int): Int = {
            val data = ByteBuffer.allocate(len)
            val n = channel.read(data)
            if (n >= 0) {
              data.position(0)
              data.get(b, off, n)
              n
            } else n
          }
        }

      case _ =>
        val is = java.lang.System.in

        new InputStream {
          override def read(): Int =
            sys.error("Unexpected operation performed by BufferedReader")
          override def available(): Int =
            is.available()

          override def read(b: Array[Byte], off: Int, len: Int): Int = {
            @tailrec def go(): Int =
              if (is.available() > 0) {
                is.read(b, off, len)
              } else {
                Thread.sleep(500L)
                go()
              }

            go()
          }
        }
    }

    val reader = new BufferedReader(new InputStreamReader(interruptibleInputStream))
    () => reader.readLine()
  }

  @silent final val getStrLn =
    zio.blocking.effectBlockingInterrupt { interruptibleStdInReader()() }.orDie

  import zio.duration._

  override def run(args: List[String]): ZIO[zio.ZEnv, Nothing, Int] = {
    import zio.console.putStrLn
    val a = getStrLn.flatMap(putStrLn(_))
    val b = ZIO.sleep(10.second) *> putStrLn("No input")
    (a race b).as(0)
  }
```

</details>

**Comment by jdegoes** [comment]
@alexknvl This is really great! I'd shy away from reflection if another solution is possible. Does your alternate formulation still support interruption?

**Comment by alexknvl** [comment]
@jdegoes https://github.com/alexknvl/interruptible-getstrln
All three strategies work equally well as far as I can tell. All three versions are interruptible.

**Comment by domdorn** [comment]
I just made this and its working for me.. as I'm new to ZIO it might have some problems? 
```
package com.dominikdorn.cb
import java.io.{BufferedReader, IOException, PrintStream}

import zio._
import zio.clock.Clock

import scala.io.StdIn
import scala.util.Try
import scala.{Console => SConsole}

class InterruptibleConsole(clock: Clock.Service)
    extends zio.console.Console.Service {

  /**
    * Prints text to the console.
    */
  def putStr(line: String): UIO[Unit] =
    putStr(SConsole.out)(line)

  /**
    * Prints text to the standard error console.
    */
  override def putStrErr(line: String): UIO[Unit] =
    putStrLn(SConsole.err)(line)

  final def putStr(stream: PrintStream)(line: String): UIO[Unit] =
    IO.effectTotal(SConsole.withOut(stream) {
      SConsole.print(line)
    })

  /**
    * Prints a line of text to the standard error console, including a newline character.
    */
  override def putStrLnErr(line: String): UIO[Unit] =
    putStrLn(SConsole.err)(line)

  /**
    * Prints a line of text to the console, including a newline character.
    */
  override def putStrLn(line: String): UIO[Unit] =
    putStrLn(SConsole.out)(line)

  final def putStrLn(stream: PrintStream)(line: String): UIO[Unit] =
    IO.effectTotal(SConsole.withOut(stream) {
      SConsole.println(line)
    })

  /**
    * Retrieves a line of input from the console.
    */
  final val getStrLn: IO[IOException, String] =
    getStrLn(SConsole.in)

  /**
    * Retrieves a line of input from the console.
    * Fails with an [[java.io.EOFException]] when the underlying [[java.io.Reader]]
    * returns null.
    */
  final def getStrLn(reader: BufferedReader): IO[IOException, String] = {
    import zio.duration._
    val v: ZIO[Clock, IOException, String] = ZIO
      .sleep(200.millis)
      .repeatUntil(_ => reader.ready()) *> (ZIO
      .fromTry(Try(StdIn.readLine())))
      .refineOrDie[IOException] {
        case e: IOException => e
      }
    v.provide(Has(clock))
  }

}

```

I can use it like this in my code:
```
    val consoleToUse2 = ZIO.access[Clock](c => new InterruptibleConsole(c.get): Console.Service).toLayer
    val consoleToUse: ZLayer[Any, Nothing, Console] = Clock.live >>> consoleToUse2
...
        input <- ZIO.sleep(4.seconds).map(_ => "4 seconds have passed").race(console.getStrLn)
...

```

**Comment by domdorn** [comment]
so I've thought a little bit more about this. 
If we'd role our own subclass of BufferedReader and add a callback here (see screenshot line 189), we could make the whole thing non-blocking without calls to sleep or any other active polling. 
<img width="674" alt="grafik" src="https://user-images.githubusercontent.com/100349/99914745-c1860300-2cff-11eb-82ce-1ce4c91b2e5f.png">

**Comment by quelgar** [comment]
> No reflection necessary, but unclear how this will interact with System.in's FileChannel

A `FileInputStream`s channel is only instantiated if requested. Actually requesting the channel of `System.in` requires shenanigans like what JRuby is doing, so I'd expect that under normal conditions `System.in` wouldn't have a channel allocated.

One possibility is to try the reflection hack to get the `System.in` channel, and if that fails due to JVM differences or whatever, fallback to making our own `FileChannel` from `FileDescriptor.in`.

**Comment by aartigao** [comment]
@alexknvl What's the status of your POC? I don't see a PR for that, but you did invest some time to fix this and #3417 🤔 

Was it discarded for some reason not specified here? Thanks.
